File: Exercises/Very Simple Simulator/src/very_simple_simulator.cpp
#include <iostream>
#include <vector>
#include <algorithm>
#include <fstream>
#include <sstream>

using namespace std;

// Define the Task structure
struct Task {
    string id;
    int BCET;
    int WCET;
    int period;  
    int deadline; 
    int priority; // Lower number => higher priority for RMS, if you wish
};

// Define the Job structure
struct Job {
    Task task;
    int releaseTime;
    int remainingTime;
    int responseTime;
};

// Function to read tasks from a CSV file
void readTasksCSV(const string& filename, vector<Task>& tasks)
{
    ifstream file(filename);
    if (!file.is_open()) 
    {
        cerr << "Error: Could not open file " << filename << endl;
        return;
    }

    string line;
    getline(file, line); // Skip header

    while (getline(file, line))
    {
        stringstream ss(line);
        Task task;
        string token;

        getline(ss, token, ','); // Task ID
        task.id = token;
        getline(ss, token, ','); // WCET
        task.WCET = stoi(token);
        getline(ss, token, ','); // BCET
        task.BCET = stoi(token);
        getline(ss, token, ','); // Period
        task.period = stoi(token);
        getline(ss, token, ','); // Deadline
        task.deadline = stoi(token);
        getline(ss, token, ','); // Priority
        task.priority = stoi(token);

        tasks.push_back(task);
    }
    file.close();
}

// Function to check if all jobs are completed
bool allJobsCompleted(const vector<Job>& jobs)
{
    for (const auto& job : jobs) {
        if (job.remainingTime > 0) {
            return false;
        }
    }
    return true;
}

// Function to find the job with the highest priority
Job *highest_priority(vector<Job *>& readyList)
{
    return *min_element(
        readyList.begin(),
        readyList.end(),
        [](Job* a, Job* b) {
            return a->task.priority < b->task.priority; 
        }
    );
}

// Function to get the list of ready jobs
vector<Job*> get_ready(vector<Job>& jobs, int currentTime)
{
    vector<Job*> readyList;
    for (auto& j : jobs) {
        if (j.releaseTime <= currentTime && j.remainingTime > 0) {
            readyList.push_back(&j);
        }
    }
    return readyList;
}

int main(int argc, char* argv[])
{
    if (argc != 2) {
        cerr << "Usage: " << argv[0] << " <tasks.csv>" << endl;
        return 1;
    }

    string filename = argv[1];
    vector<Task> tasks;
    readTasksCSV(filename, tasks);

    int n = 1000; 
    int currentTime = 0;

    vector<Job> jobs;
    jobs.reserve(tasks.size());
    for (const auto& t : tasks) {
        jobs.push_back({t, 0, t.WCET, 0});
    }

    vector<int> worstCaseResponseTimes(tasks.size(), 0);

    while (currentTime < n && !allJobsCompleted(jobs))
    {
        // Release new jobs at the start of each period
        for (const auto& t : tasks) 
        {
            if (currentTime != 0 && currentTime % t.period == 0) {
                jobs.push_back({t, currentTime, t.WCET, 0});
            }
        }

        // Get the list of ready jobs
        vector<Job*> readyList = get_ready(jobs, currentTime);

        // If no jobs are ready, increment the current time
        if (readyList.empty()) 
        {
            currentTime++;
            continue;
        }

        // Get the job with the highest priority
        Job* currentJob = highest_priority(readyList);

        if (currentJob != nullptr)
        {
            // Execute the job for one time unit
            currentJob->remainingTime--;
            currentTime++;
    
            // If the job is completed, update its response time
            if (currentJob->remainingTime <= 0) 
            {
                currentJob->responseTime = currentTime - currentJob->releaseTime;
                int taskIndex = distance(
                    tasks.begin(),
                    find_if(tasks.begin(), tasks.end(), [&](const Task& tk){
                        return tk.id == currentJob->task.id;
                    })
                );
                worstCaseResponseTimes[taskIndex] = max(
                    worstCaseResponseTimes[taskIndex],
                    currentJob->responseTime
                );
            }
        }
        else
        {
            currentJob->remainingTime--;
            currentTime++;
        }
    }

    // Output the worst-case response times for each task
    for (size_t i = 0; i < tasks.size(); ++i) 
    {
        cout << "Task " << tasks[i].id
             << " WCRT: " << worstCaseResponseTimes[i] << endl;
    }

    return 0;
}

File: Exercises/Very Simple Simulator/src/very_simple_simulator.py
import pandas as pd
import numpy as np
from dataclasses import dataclass


# struct to keep track of the jobs
@dataclass
class Jobs:
    task: np.array
    release_time: int
    remaining_time: int
    response_time: int


# function to read the file
def read_file(path: str) -> pd.DataFrame:
    """
    Reads a CSV file from the specified path and returns its contents as a pandas DataFrame.
    Parameters:
    path (str): The file path to the CSV file.
    Returns:
    pd.DataFrame: The contents of the CSV file as a pandas DataFrame.
    """
    return pd.read_csv(path)


# function to check if every job is done
def check_remaining_jobs(jobs: np.array) -> bool:
    """
    Checks if there are any jobs remaining to be executed.
    Parameters:
    jobs (np.array): The list of jobs to check.
    Returns:
    bool: True if there are jobs remaining, False otherwise.
    """
    for job in jobs:
        if job.remaining_time > 0:
            return True
    return False


# function to check if the job is ready to be executed
def get_ready_list(jobs: np.array, current_time: int) -> np.array:
    ready_job: np.array = []
    for job in jobs:
        if job.release_time <= current_time and job.remaining_time > 0:
            ready_job.append(job)
    return ready_job


def run_cycle(input_path: str) -> None:
    # Print the contents of the CSV file
    # Task  BCET(best case execution time)  WCET(worst case execution time)  Period  Deadline  Priority
    df = read_file(input_path).values

    # init number of cycles and current time
    num_cycles: int = 1000
    current_time: int = 0

    # init the jobs
    jobs: np.array = []
    for i in range(len(df)):
        jobs.append(Jobs(df[i], 0, df[i, 2], 0))

    jobs_queue = jobs.copy()

    # init the worst case dictionary
    wcrt_dict = {jobs[i].task[0]: 0 for i in range(len(jobs))}

    while current_time <= num_cycles and check_remaining_jobs(jobs_queue):
        # release job at the start of the period
        for job in jobs:
            if current_time % job.task[3] == 0 and current_time != 0:
                jobs_queue.append(Jobs(job.task, current_time, job.task[2], 0))

        ready_job = get_ready_list(jobs_queue, current_time)
        highest_priority_job = None
        if ready_job:
            highest_priority_job = max(ready_job, key=lambda job: job.task[5])

        if highest_priority_job:
            # advace by 1
            current_time += 1
            # compute the task
            highest_priority_job.remaining_time -= 1
            # check if the task is done
            if highest_priority_job.remaining_time <= 0:
                highest_priority_job.response_time = (
                    current_time - highest_priority_job.release_time
                )
                wcrt_dict[highest_priority_job.task[0]] = max(
                    wcrt_dict[highest_priority_job.task[0]],
                    highest_priority_job.response_time,
                )
        else:
            current_time += 1
    # Create a list to store the results
    results = []

    for job in jobs:
        task = job.task[0]
        wcrt = wcrt_dict[task]
        deadline = job.task[4]
        status = "true" if wcrt <= deadline else "false"
        results.append([task, wcrt, deadline, status])

    # Create a DataFrame from the results
    results_df = pd.DataFrame(results, columns=["Task", "WCRT", "Deadline", "Status"])

    # Save the DataFrame to a CSV file
    results_df.to_csv(input_path.removesuffix(".csv") + "_s.csv", index=False)


if __name__ == "__main__":
    run_cycle(
        "/Users/pierfrancesco/Desktop/second semester DTU/distributed/repo/Exercises/Very Simple Simulator/exercise-TC1.csv"
    )
